base58.go

```diff
@@ -13,45 +17,90 @@ func Decode(input string) []byte {
 		return []byte("")
 	}
 
-	// The max possible output size is when a base58 encoding consists of
-	// nothing but the alphabet character at index 0 which would result in the
-	// same number of bytes as the number of input chars.
-	output := make([]byte, len(input))
+	// Determine the maximum possible output size.
+	//
+	// Since the conversion is from base58 to base256, the max possible number
+	// of bytes of output per input byte, excluding the leading zeros, is
+	// log_256(58).  Therefore, the max total output size is the number of
+	// leading zero bytes plus ceil(inputSizeMinusLeadingZeros * log_256(58)).
+	//
+	// Note that log_256(58) ~= 0.7322 < 47/64 which is within 0.3% of the true
+	// value and efficient to compute as it only involves division by a power of
+	// 2 and thus serves as a good approximation.  So, the calculation below is
+	// the integer division equivalent of nlz + ceil(len(input[nlz:]) * 47/64).
+	//
+	// Finally, in order to avoid additional conditional branches in the
+	// conversion from uint32s to bytes, the max output size is rounded up to
+	// the next multiple of 4.
+	var nlz int
+	for i := 0; i < len(input) && input[i] == alphabetIdx0; i++ {
+		nlz++
+	}
+	maxOutputSizeNoLZ := (len(input[nlz:])*47 + 63) / 64
+	maxOutputSize := nlz + maxOutputSizeNoLZ
+	maxOutputSize = ((maxOutputSize + 3) / 4) * 4
+	output := make([]byte, maxOutputSize)
+
+	// The algorithm below performs the calculations with uint32s for better
+	// performance and the total number of uint32s is ceil(maxOutputSizeNoLZ /
+	// 4).  Note that the leading zeros are skipped here, so the calculation is
+	// based on the max output size excluding them.
+	//
+	// In order to avoid an additional heap allocation for the vast majority of
+	// typical cases, use an array on the stack for inputs of up to 120 chars
+	// (plus any leading zeros) and fall back to a heap alloc for larger inputs.
+	//
+	// This value was chosen because it provides a good balance between alloc
+	// size, speed, and the max chars in the vast majority of inputs decoded in
+	// the most common use cases.
+	const maxOut32StackAlloc = 120 / 4
```

Comment:
This can be updated to account for the resulting max output size for the target input size as opposed to just being the target input size since the former is smaller.

Reply:
Changed.